docs: fix bug in code examples for downgrading a service

bertyhell opened the pull request.

gkalpak commented:
The service is indeed called `heroes` in [AngularJS-land][1]. This change seems invalid.

[1]: https://github.com/bertyhell/angular/blob/90d0f79b9fed1435e6c97ea05fdcb62d33aea570/aio/content/examples/upgrade-module/src/app/ajs-to-a-providers/app.module.ts#L38

bertyhell commented:
> The service is indeed called heroes in AngularJS-land. This change seems invalid.

If that is the case, the documentation is even more confusing, since it states the opposite here:
https://angular.io/guide/upgrade#making-angularjs-dependencies-injectable-to-angular

> In these situations, it is possible to upgrade an AngularJS provider to Angular. This makes it possible to then inject it somewhere in Angular code. For example, you might have a service called HeroesService in AngularJS:

gkalpak commented:
It is indeed somewhat confusing, because we never show how the service is registered in AngularJS. In both Angular and AngularJS there are two things that identify a registered service:
1. The injection token, i.e. an identifier that we will use to request the service from the injector.
2. The service class (which will be used by the injector to create an instance of the service).

In Angular, it is common to use the class itself as the injection token:

```ts
providers: [
  {provide: MyClass, useClass: MyClass},
  // or the shorthand:
  MyClass
]
```

This is possible, because Angular's injection system can accept any token as identifier.

In AngularJS on the other hand, we only accept string identifiers as token. Thus you are forced to use two different things: a string (for the injection token) and a class (for the service)

```ts
.service('myServiceId', MyServiceClass)
```

It is common in AngularJS apps to chose a service name for the token (e.g. `user`) and append the `Service` suffix to create the class name:

```ts
.service('user', class UserService { ... })
```

##
Hopefully that clears it up a bit. Feel free to update to clarify things (as long as the change is valid :smiley:).

bertyhell commented:
Ok, I changed the token name back to the original and added a note to explain the 'heroes' string.
Does that work for you?

https://github.com/angular/angular/pull/19371/files#diff-33c8eb0a067223c3fbf3dd45ec689efdR783

trotyl commented:
> This is possible, because Angular's injection system can accept any token as identifier.

Hi @gkalpak , sorry for off topic, I guess there're some different voices on https://github.com/angular/angular/pull/15319#issuecomment-287965852, may I know what is the expected validity for DI token now (or planned future)?

bertyhell commented:
@trotyl Good point, the question is: After the 5.0 release will people be upgrading their app from 1.x => 4.x or 1.x => 5.x.

Either way, the docs will have to be updated to reflect this, but that's beyond the scope of this PR.

gkalpak commented:
@trotyl, yes, things might change in future version. I am not sure what the current state is.
@bertyhell, what we have in the docs now will work with 4.x and 5.x alike, since we are using a `Type` (`HeroesService`) as token (in this case at least).

gkalpak commented on aio/content/examples/upgrade-module/src/app/ajs-to-a-providers/hero-detail.component.ts:
```diff
@@ -5,6 +5,7 @@ import { Hero } from '../hero';
 
 @Component({
   selector: 'hero-detail',
+  providers: [HeroesService],
```
Why this change?

bertyhell replied:
@gkalpak Without this change I get this error in the app i'm trying to migrate:

```
ERROR Error: Uncaught (in promise): Error: [$injector:unpr] Unknown provider: heroesServiceProvider <- heroesService
```

gkalpak replied:
This example provides the service on the `NgModule`. Providing it on the component too, will create a separate instance for the component (different than what the rest of the app is using). This is not what we want here (but might be what you want in your app).

To share the same instance of the service across components, put it on the `providers` array of the `NgModule`.

bertyhell commented:
> To share the same instance of the service across components, put it on the providers array of the NgModule.

I modified the code and added some more description text to the readme.

kapunahelewong commented on aio/content/guide/upgrade.md:
```diff
@@ -778,16 +778,25 @@ delete them once the upgrade is over.
+
+Note that the 'heroes' string inside the factory refers to the AngularJS `HeroesService`.
+It is common in AngularJS apps to chose a service name for the token (e.g. heroes)
+and append the Service suffix to create the class name (e.g. heroesService)
```
Could you please change "class name (e.g. heroesService)" to "class name, for example `heroesService`"? The reasons for changes are:

1. We are trying to minimize parentheticals whenever we can.
2. We're replacing i.e. and e.g. with "that is" and "for example" since people often confuse them.
3. For consistency, `heroesService` should be in code font, thus the back ticks.

kapunahelewong commented on aio/content/guide/upgrade.md:
```diff
@@ -778,16 +778,25 @@ delete them once the upgrade is over.
 </code-example>
 
-You can then inject it in Angular using it's class as a type annotation:
+Then use the service inside your component by injecting it in the component constructor using it's class as a type annotation:
```
it's --> its

I know this was already there, but since you're on the line, would you mind deleting the apostrophe?

kapunahelewong reviewed (changes requested):
Thanks for the updates @bertyhell. Just a couple of comments for you.

bertyhell commented:
@kapunahelewong Sorry for the long delay. I missed the email notification about your comments.

kapunahelewong commented:
Hey @bertyhell, if you'd like to rebase, I'd be happy to approve :)

bertyhell committed:
docs: Improve docs for downgrading a service

brandonroberts commented:
@gkalpak are there any changes left on this from your point of view?

brandonroberts commented on aio/content/guide/upgrade.md:
```diff
@@ -778,16 +778,25 @@ delete them once the upgrade is over.
 It's also recommended to export the `heroesServiceFactory` function so that Ahead-of-Time
 compilation can pick it up.
 
+<div class="l-sub-section">
```
Change `l-sub-section` to `alert is-helpful`

brandonroberts reviewed (changes requested):
Minor css change

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -778,16 +778,25 @@ delete them once the upgrade is over.
+<div class="l-sub-section">
+
+Note that the 'heroes' string inside the factory refers to the AngularJS `HeroesService`.
+It is common in AngularJS apps to chose a service name for the token, for example heroes,
```
_chose_ --> _choose_
_for example heroes_ --> _for example "heroes"_

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -778,16 +778,25 @@ delete them once the upgrade is over.
+
+Note that the 'heroes' string inside the factory refers to the AngularJS `HeroesService`.
+It is common in AngularJS apps to chose a service name for the token, for example heroes,
+and append the Service suffix to create the class name, for example `heroesService`.
```
_Service_ --> _"Service"_
Also, classes (or constructor functions) start with an uppercase letter (by convention). So `HeroesService`, not `heroesService`.

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -778,16 +778,25 @@ delete them once the upgrade is over.
 <code-example path="upgrade-module/src/app/ajs-to-a-providers/ajs-upgraded-providers.ts" title="ajs-upgraded-providers.ts">
 </code-example>
 
+You can then add the service to Angular by adding it to the `@NgModule`:
```
I would change "add the service" to "provide the service".

gkalpak reviewed (commented):
a couple of minor comments. LGTM otherwise.

brandonroberts commented:
@gkalpak this one is ready for another review

gkalpak reviewed (approved):
https://github.com/angular/angular/pull/19371#discussion_r207909524 hasn't been fully addressed. Other than that LGTM.
